Use shape for DataIO length

ajtritt opened the pull request:
## Motivation

When shape is specified in DataIO, length should be obtainable. However, because `data` was not being set, DataIO was declaring itself not valid and raising an error.

This also adds the requirement in H5DataIO that `dtype` and `shape` are specified together. [h5py's `create_dataset`](https://docs.h5py.org/en/stable/high/group.html#h5py.Group.create_dataset) sets `dtype` to 32-bit float by default. I opted to not rely on this default behavior out of concern of creating unintended behavior in hdmf. 

## How to test the behavior?
```python
dataio = H5DataIO(dtype=int, shape=(10, 10))
assert len(dataio) == 10
```

## Checklist

- [ ] Did you update CHANGELOG.md with your changes?
- [ ] Have you checked our [Contributing](https://github.com/hdmf-dev/hdmf/blob/dev/docs/CONTRIBUTING.rst) document?
- [ ] Have you ensured the PR clearly describes the problem and the solution?
- [ ] Is your contribution compliant with our coding style? This can be checked running `flake8` from the source directory.
- [ ] Have you checked to ensure that there aren't other open [Pull Requests](https://github.com/hdmf-dev/hdmf/pulls) for the same change?
- [ ] Have you included the relevant issue number using "Fix #XXX" notation where XXX is the issue number? By including "Fix #XXX" you allow GitHub to close issue #XXX when the PR is merged.

ajtritt committed:
Use class from predefined column

- Previously, the 'class' key of column configs for autogenerated
  columns was not being set, which led to setting this key to *VectorData*
  by default. This change uses the data_type_inc from a DynamicTable
  predefined column to set the 'class' key in the autogenerated.

ajtritt committed:
Use shape determine length of DataIO/H5DataIO

- Also add tests

ajtritt committed:
Use shape or maxshape to determine length of DataIO/H5DataIO

ajtritt committed:
update changelog

ajtritt committed:
Merge branch 'enh/h5dataio_len' of github.com:hdmf-dev/hdmf into enh/h5dataio_len

rly commented on src/hdmf/data_utils.py:
```diff
@@ -995,6 +995,8 @@ def __deepcopy__(self, memo):
     def __len__(self):
         """Number of values in self.data"""
+        if self.__shape is not None:
+            return self.__shape[0]
```
Is it possible for `self.__shape` to become out of sync with `self.data`? For example, if the user creates a DataIO without data and then sets data later but sets it to a shape not equal to self.__shape (or a dtype not equal to self.__dtype)?

ajtritt replied:
If I understand L950 correctly, no that's not possible.

ajtritt replied:
And I just added a test to make sure that doesn't happen.

rly replied:
Ah, I see. Looks good. Thanks!

ajtritt committed:
remove unnecessary code

rly commented on src/hdmf/common/io/table.py:
```diff
@@ -82,6 +82,8 @@ def process_field_spec(cls, classdict, docval_args, parent_cls, attr_name, not_i
             # the user must specify the table attribute on the DTR after it is generated
             column_conf['table'] = True
+        else:
+            column_conf['class'] = dtype
```
I *think* this flow would work for subclasses of `VectorData` but it would be good to test that.

ajtritt replied:
Do you mean for stuff like `EnumData`, `DynamicTableRegion`, etc?

rly replied:
yeah, well not `DynamicTableRegion`, because that's handled above, and I am not sure `EnumData` is fully handled by `__columns__` but I think it is. I was more thinking of custom extensions to VectorData. Making such a test might be a pain though. I think it would be fine but feel free to add a test.

rly commented:
Besides the comment above, this looks good to me.

ajtritt committed:
add test for shape/data check

oruebel commented on src/hdmf/backends/hdf5/h5_utils.py:
```diff
@@ -489,6 +489,10 @@ def __init__(self, **kwargs):
         ioarg_names = [name for name in kwargs.keys() if name not in ['data', 'link_data', 'allow_plugin_filters',
                                                                       'dtype', 'shape']]
+
+        if kwargs['data'] is None:
+            if kwargs['shape'] is None or kwargs['dtype'] is None:
+                raise ValueError("Must specify 'dtype' and 'shape' if not specifying 'data'")
```
Should this check be part of ``DataIO.__init__`` instead? I can't think of a use case where we want to allow this in ``DataIO`` (or any subclass thereof) but not in ``HDF5IO``.

ajtritt replied:
Yes, it has been moved to DataIO

ajtritt committed:
Rearrange data vs dtype+shape code

- Move check for dtype+shape when data not specified to DataIO
- Issue warnings if dtype or shape passed if data is specified. Before
  this raised an error, but now they will be ignored i.e. set to None

ajtritt committed:
Update tests to check for warnings

ajtritt committed:
Make dtype+shape check exclusive or

- This was required for backwards compatability i.e. allow setting
  DataIO.data with the setter and not in the constructor call

oruebel commented on src/hdmf/data_utils.py:
```diff
@@ -922,7 +922,7 @@ class DataIO:
         data, dtype, shape = popargs('data', 'dtype', 'shape', kwargs)
         if data is None:
-            if dtype is None or shape is None:
+            if (dtype is None) ^ (shape is None):
```
Why use XOR instead of OR here? Should this error not also be raised if both dtype and shape are None?

ajtritt replied:
See the commit message. It's done for backward compatibility i.e. set_dataio
